src/esp32/encoder.rs

```diff
@@ -0,0 +1,282 @@
+                    Ok(num) => num,
+                    Err(_) => return Err(anyhow::anyhow!("cannot build encoder, need 'a' pin")),
+                };
+                let pin_b_num = match cfg.get_attribute::<i32>("b") {
+                    Ok(num) => num,
+                    Err(_) => return Err(anyhow::anyhow!("cannot build encoder, need 'b' pin")),
+                };
+                let a = match PinDriver::input(unsafe { AnyInputPin::new(pin_a_num) }) {
+                    Ok(a) => a,
+                    Err(err) => {
+                        return Err(anyhow::anyhow!(
+                            "cannot build encoder, could not initialize pin {:?} as pin 'a': {:?}",
+                            pin_a_num,
+                            err
+                        ))
+                    }
+                };
+                let b = match PinDriver::input(unsafe { AnyInputPin::new(pin_b_num) }) {
+                    Ok(b) => b,
+                    Err(err) => {
+                        return Err(anyhow::anyhow!(
+                            "cannot build encoder, could not initialize pin {:?} as pin 'b': {:?}",
+                            pin_b_num,
+                            err
+                        ))
+                    }
+                };
+                Ok(Arc::new(Mutex::new(Esp32Encoder::new(a, b)?)))
+            }
+        }
+    }
+
+    pub fn start(&self) -> anyhow::Result<()> {
+        unsafe {
+            match esp_idf_sys::pcnt_counter_resume(self.config.unit) {
+                ESP_OK => {}
+                err => return Err(EspError::from(err).unwrap().into()),
+            }
+        }
+        Ok(())
+    }
+    pub fn stop(&self) -> anyhow::Result<()> {
+        unsafe {
+            match esp_idf_sys::pcnt_counter_pause(self.config.unit) {
+                ESP_OK => {}
+                err => return Err(EspError::from(err).unwrap().into()),
+            }
+        }
+        Ok(())
+    }
+    pub fn reset(&self) -> anyhow::Result<()> {
+        self.stop()?;
+        unsafe {
+            match esp_idf_sys::pcnt_counter_clear(self.config.unit) {
+                ESP_OK => {}
+                err => return Err(EspError::from(err).unwrap().into()),
+            }
+        }
+        self.start()?;
+        Ok(())
+    }
+    pub fn get_counter_value(&self) -> anyhow::Result<i32> {
+        let mut ctr: i16 = 0;
+        unsafe {
+            match esp_idf_sys::pcnt_get_counter_value(self.config.unit, &mut ctr as *mut c_short) {
+                ESP_OK => {}
+                err => return Err(EspError::from(err).unwrap().into()),
+            }
+        }
+        let tot = self.pulse_counter.acc.load(Ordering::Relaxed) * 100 + i32::from(ctr);
+        Ok(tot)
+    }
+    pub fn setup_pcnt(&mut self) -> anyhow::Result<()> {
+        unsafe {
+            match esp_idf_sys::pcnt_unit_config(&self.config as *const pcnt_config_t) {
+                ESP_OK | ESP_ERR_INVALID_STATE => {}
```

Comment:
remove ESP_ERR_INVALID_STATE it's a mistake from my side it shouldn't be considered as a success